listenbrainz_spark/stats/sitewide/listening_activity.py

```diff
@@ -0,0 +1,103 @@
+import json
+import logging
+from datetime import datetime
+from typing import Iterator, Optional
+
+from pydantic import ValidationError
+
+from data.model.user_listening_activity import UserListeningActivityStatMessage, SitewideListeningActivityStatMessage
+from listenbrainz_spark.stats import run_query
+from listenbrainz_spark.stats.common.listening_activity import setup_time_range
+from listenbrainz_spark.utils import get_listens_from_new_dump
+from pyspark.sql.types import (StringType, StructField, StructType,
+                               TimestampType)
+
+time_range_schema = StructType([
+    StructField("time_range", StringType()),
+    StructField("start", TimestampType()),
+    StructField("end", TimestampType())
+])
+
+
+logger = logging.getLogger(__name__)
+
+
+def calculate_listening_activity():
+    """ Calculate number of listens for each user in time ranges given in the "time_range" table.
+    The time ranges are as follows:
+        1) week - each day with weekday name of the past 2 weeks.
+        2) month - each day the past 2 months.
+        3) year - each month of the past 2 years.
+        4) all_time - each year starting from LAST_FM_FOUNDING_YEAR (2002)
+    """
+    # calculates the number of listens in each time range for each user, count(listen.listened_at) so that
+    # group without listens are counted as 0, count(*) gives 1.
+    result = run_query(""" 
```

Comment:
I like how you're using the time_range table here -- clever!

Reply:
😄